src/librustc_typeck/check/mod.rs

```diff
@@ -1802,8 +1803,36 @@ fn check_transparent<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, sp: Span, def_id: De
         return;
     }
 
+    if adt.is_enum() {
+        if !tcx.features().transparent_enums {
+            emit_feature_err(&tcx.sess.parse_sess,
+                             "transparent_enums",
+                             sp,
+                             GateIssue::Language,
+                             "transparent enums are unstable");
+        }
+        if adt.variants.len() != 1 {
+            let variant_spans: Vec<_> = adt.variants.iter().map(|variant| {
+                tcx.hir().span_if_local(variant.def_id).unwrap()
+            }).collect();
+            struct_span_err!(tcx.sess, sp, E0726,
+                            "transparent enum needs exactly one variant, but has {}",
+                            adt.variants.len())
+            .span_note(variant_spans, "variant")
```

Comment:
I don't think this really adds anything. Also, it doesn't work properly when there are zero variants.

Reply:
> I don't think this really adds anything.

I (respectfully) disagree. I've been using Rust for awhile now (I'm not an expert but I'm not a noob either) and I only learned what "variant" meant while writing this pull request. If I had the error message `"transparent enum needs exactly one variant"` I would have to pause for a bit to figure out what a "variant" was (I probably could have figured it out, but the `"but has N"` message with the variant spans highlighted would have helped). It helps clarify some of the jargon.

I also like that it follows Rust's usual error message pattern of `"needs/expected X, but got Y"`.

> Also, it doesn't work properly when there are zero variants.

[I think it works properly](https://github.com/rust-lang/rust/blob/333b63659d650fc9a46bed508455371cc6454958/src/test/ui/repr/repr-transparent.rs#L44). Can you please clarify?

Reply:
> It helps clarify some of the jargon.

Maybe we could just change the note to something a bit more descriptive, then: e.g. "these variants exist on `Foo`" (or similar).

> Can you please clarify?

[What I mean is this:](https://github.com/rust-lang/rust/pull/60463/files#diff-e2193f50e3676dacb00aadfd2cffa893R83) it just says "variant", which doesn't make any sense when there aren't any. It should be special cased to not show this note in that case.

Reply:
> [What I mean is this:](https://github.com/rust-lang/rust/pull/60463/files#diff-e2193f50e3676dacb00aadfd2cffa893R83) it just says "variant"

Thanks! [That made me realize the existing error message for transparent `struct`s has a similar issue](https://rust.godbolt.org/z/LBUdhq). I've revised the code to only emit the note when more than 1 non-ZST field is present (and changed "non-zero-sized field" to the plural form "non-zero-sized fields").